enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,622 @@
+> [Management Workload Partitioning](management-workload-partitioning.md)
+
+- This enhancement is focused on CPU resources. Other compressible resource
+  types may need to be managed in the future, and those are likely to need
+  different approaches.
+- This enhancement does not address non-compressible resource requests, such as
+  for memory.
+- This enhancement does not address ways to disable operators or operands
+  entirely.
+- This enhancement does not address reducing actual utilization, beyond
+  providing a way to have a predictable upper-bounds. There is no expectation
+  that a cluster configured to use a small number of cores for management
+  services would offer exactly the same performance as the default. It must be
+  stable and continue to operate reliably, but may respond more slowly.
+- **This enhancement does not address mixing nodes with pinning and without,
+  this feature will be enabled cluster wide and encapsulate both master and
+  worker pools. If it's not desired then the setting will still be turned on but
+  the management workloads will run on the whole CPU set for that desired
+  pool.**
+- **This enhancement assumes that the configuration of a management CPU pool is
+  done as part of installing the cluster. It can be changed after the fact but
+  we will need to stipulate that, that is currently not supported. The intent
+  here is for this to be supported as a day 0 feature, only.**
+
+## Proposal
+
+We will need to maintain a global identifier that is set during installation and
+can not be easily removed after the fact. This approach will help remove
+exposing this feature via an API and limiting the chances that a
+misconfiguration can cause un-recoverable scenarios for our customers. At
+install time we will also apply an initial machine config for workload
+partitioning that sets a default CPUSet for the whole CPUSet. Effectively this
+will behave as if workload partitioning is not turned on. With this approach we
+eliminate the race condition that can occur if we apply a machine config after
+the fact as nodes join the cluster. When a customer wishes to pin the management
+workloads they will be able to do that via the existing Performance Profile.
+Resizing partition size will not cause any issues after installation.
+
+In order to implement this enhancement we are proposing changing 4 components
+defined below.
+
+1. Openshift API - ([Infrastructure
+   Status](https://github.com/openshift/api/blob/81fadf1ca0981f800029bc5e2fe2dc7f47ce698b/config/v1/types_infrastructure.go#L51))
+
+   - The change in this component is to store a global identifier if we have
+     partitioning turned on.
+
+2. [Openshift Installer](https://github.com/openshift/installer))
+
+   - The change in the installer is to support turning on this feature only
+     during installation as well as apply the partitioning machine configs so
+     that we avoid the race conditions when running in multi-node environments.
```

Comment:
That's a great call out, I might need to re-work the wording here a bit, essentially there are two explicit intents that I'm trying to capture. Which should help answer those questions.

The first intent is enabling the feature for the cluster. In regard to kubelet, there are technically two configuration files to be aware of, the one we care about here is the `/etc/kubernetes/openshift-workload-pinning` which makes kubelet aware of managed pods. So the installer will update the infrastructure resource with the partitioning status to simply denoting that this cluster has been configured with cpu partitioning enabled but no workload pinning is applied, so at this stage even the kubelet process itself isn't pinned since that level of specificity will be driven by the NTO via the Performance Profile after the bootstrap.

The second intent is pinning specific workloads to specific CPU sets. This is where the NTO comes in, from this stage on a customer can now choose to pin specific workloads, up until now they have a cluster that has CPU Partitioning turned on, but no workloads are pinned and they must define a Performance Profile with a specific "workloads" to pin, in this enhancement we only support the "Infrastructure" workloads. Once that is applied new Machine Configs are generated and they then update those previous files to set the CPU sets as well as set the kubelet config itself for the process.

Regarding your last comment, I'm not entirely sure how MCO handles those situations. If the rendered configs are not ready from the NTO by the time the worker nodes are ready, then I'm in agreement we will need to be able to tolerate and extra reboot during that initial setup.

Reply:
The API server comes into play, too. When workload partitioning is enabled we use special resources to manage scheduling control plane workloads so we don't consume the apparent CPU capacity of the node. If the pod definition is mutated to change the normal CPU and memory requests into requests for those special resources before all of the kubelets advertise that same special resource (because they aren't already configured to do workload partitioning, yet), then the pod will be unschedulable. If that happens to the networking pods or another daemon set that is integral to making the cluster work then the node will be broken. So, when partitioning is enabled, we need to make sure it's enabled in all of the kubelets before (or at the same time as) the API server that is mutating the pods.

As long as the API server and kubelet consider the feature enabled at the same time, it doesn't really seem to matter what the CPUset is on any given node, and I think it would be safe to reconfigure the CPUset day 2.

So, let's figure out which component is responsible for ensuring that `openshift-workload-pinning` file is created for each kubelet on the first boot of each node when the user enables the feature in the installer. When we have that worked out, I think we'll have this solved.